[CSS ZOOM] Apply zoom factor to border width

nullhook opened the pull request:
#### c38095edbbdb85138b057e0099901777e0f8bf40
<pre>
[CSS ZOOM] Apply zoom factor to border width
<a href="https://bugs.webkit.org/show_bug.cgi?id=299885">https://bugs.webkit.org/show_bug.cgi?id=299885</a>
<a href="https://rdar.apple.com/161669263">rdar://161669263</a>

Reviewed by Antti Koivisto and Sam Weinig.

We apply the zoom factor at evaluation-time for LineWidth,
BorderImageWidth, and MaskBorderWidth.

Tests: imported/w3c/web-platform-tests/css/css-viewport/zoom/border-image-width-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/border-image-width.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/mask-border-width-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/mask-border-width.html

* LayoutTests/TestExpectations:

* LayoutTests/fast/sub-pixel/zoomed-em-border.html:
* LayoutTests/svg/zoom/page/zoom-replaced-intrinsic-ratio-001.htm:

We&apos;re disabling evaluation-time zoom flag on above tests because of double zoom.
The issue is that font-dependent units are being affected by the zoom
value during style-resolution, and we apply zoom again during evaluation.
This will be addressed separately as a follow-up:
<a href="https://bugs.webkit.org/show_bug.cgi?id=299816">https://bugs.webkit.org/show_bug.cgi?id=299816</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-image-width-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-image-width-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-image-width.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-computed-zoom-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-computed-zoom.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-svg-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/mask-border-width-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/mask-border-width-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/mask-border-width.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/animations/border-width-interpolation-expected.txt:

A progression

* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
(WebCore::Layout::FormattingGeometry::computedBorder const):
* Source/WebCore/layout/formattingContexts/block/BlockFormattingGeometry.cpp:
(WebCore::Layout::BlockFormattingGeometry::intrinsicWidthConstraints const):
* Source/WebCore/layout/integration/LayoutIntegrationBoxGeometryUpdater.cpp:
(WebCore::LayoutIntegration::BoxGeometryUpdater::logicalBorder):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::fixedContainerEdges const):
* Source/WebCore/page/SpatialNavigation.cpp:
(WebCore::nodeRectInAbsoluteCoordinates):
* Source/WebCore/rendering/BorderEdge.cpp:
(WebCore::borderEdges):
(WebCore::borderEdgesForOutline):
* Source/WebCore/rendering/BorderPainter.cpp:
(WebCore::BorderPainter::paintOutline const):
* Source/WebCore/rendering/BorderShape.cpp:
(WebCore::BorderShape::shapeForBorderRect):
* Source/WebCore/rendering/NinePieceImagePainter.cpp:
(WebCore::computeSlice):
(WebCore::computeSlices):
(WebCore::paintNinePieceImage):
* Source/WebCore/rendering/RenderBoxModelObjectInlines.h:
(WebCore::RenderBoxModelObject::borderAfter const):
(WebCore::RenderBoxModelObject::borderBefore const):
(WebCore::RenderBoxModelObject::borderBottom const):
(WebCore::RenderBoxModelObject::borderEnd const):
(WebCore::RenderBoxModelObject::borderLeft const):
(WebCore::RenderBoxModelObject::borderRight const):
(WebCore::RenderBoxModelObject::borderStart const):
(WebCore::RenderBoxModelObject::borderTop const):
(WebCore::RenderBoxModelObject::borderWidths const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::drawFocusRing):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintAreaElementFocusRing):
* Source/WebCore/rendering/RenderMultiColumnSet.cpp:
(WebCore::RenderMultiColumnSet::paintColumnRules):
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::calcBorderStart const):
(WebCore::RenderTable::calcBorderEnd const):
(WebCore::RenderTable::outerBorderBefore const):
(WebCore::RenderTable::outerBorderAfter const):
(WebCore::RenderTable::outerBorderStart const):
(WebCore::RenderTable::outerBorderEnd const):
* Source/WebCore/rendering/RenderTableCell.cpp:
(WebCore::RenderTableCell::emptyBorder const):
(WebCore::RenderTableCell::computeCollapsedStartBorder const):
(WebCore::RenderTableCell::computeCollapsedEndBorder const):
(WebCore::RenderTableCell::computeCollapsedBeforeBorder const):
(WebCore::RenderTableCell::computeCollapsedAfterBorder const):
(WebCore::emptyBorder): Deleted.
* Source/WebCore/rendering/RenderTableCell.h:
* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::calcBlockDirectionOuterBorder const):
(WebCore::RenderTableSection::calcInlineDirectionOuterBorder const):
(WebCore::RenderTableSection::paintRowGroupBorderIfRequired):
(WebCore::RenderTableSection::cachedCollapsedBorder):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::extractControlStyleForRenderer const):
(WebCore::RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle const):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::paintMenuListButtonDecorationsForVectorBasedControls):
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::popupInternalPaddingBox const):
(WebCore::RenderThemeIOS::paintMenuListButtonDecorations):
* Source/WebCore/rendering/style/CollapsedBorderValue.h:
(WebCore::CollapsedBorderValue::CollapsedBorderValue):
(WebCore::CollapsedBorderValue::m_transparent): Deleted.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setDeviceScaleFactor):
(WebCore::RenderStyle::imageOutsets const):
(WebCore::RenderStyle::outlineWidth const):
(WebCore::RenderStyle::outlineOffset const):
(WebCore::RenderStyle::outlineSize const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::deviceScaleFactor const):
(WebCore::RenderStyle::usedZoomForLength const):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::computeFloatVisibleRectInContainer):
* Source/WebCore/style/StyleResolveForDocument.cpp:
(WebCore::Style::resolveForDocument):
* Source/WebCore/style/values/backgrounds/StyleBorderImageWidth.h:
* Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:
(WebCore::Style::CSSValueConversion&lt;LineWidth&gt;::operator):
(WebCore::Style::FloatBoxExtent&gt;::operator):
(WebCore::Style::LayoutBoxExtent&gt;::operator):
* Source/WebCore/style/values/backgrounds/StyleLineWidth.h:
(WebCore::Style::LineWidth::LineWidth):
* Source/WebCore/style/values/masking/StyleMaskBorderWidth.h:
* Source/WebCore/style/values/primitives/StyleZoomPrimitives.h:
(WebCore::Style::ZoomFactor::ZoomFactor):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::absoluteInteractionBounds):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMNode innerFrameQuad]):

Canonical link: <a href="https://commits.webkit.org/301536@main">https://commits.webkit.org/301536@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46ee801328ceaef96ce38ad66b1876601977c71a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126235 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45891 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36709 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133010 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78004 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4b4a0d43-6b32-40f0-8d28-b0ad5033b28e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128106 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46552 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54434 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96106 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/64219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/fe914b4c-0267-4e1d-87b8-f72a27ac238f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129183 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37247 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112914 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76587 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2751dc69-16ca-4f30-83f0-b0b108e1e36d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36145 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31097 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76481 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107029 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31324 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135709 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52981 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40699 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/104611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53443 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109146 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104312 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26606 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49752 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28086 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50364 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52881 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58712 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52198 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55542 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53914 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9c6ecb4e92ce739017f3c11f640e3122c22f1741. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5132ae3f06eb115aca816a37cde53529034f9f27. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

nt1m commented on LayoutTests/fast/sub-pixel/zoomed-em-border.html:
```diff
@@ -1,6 +1,12 @@
     <head>
+        <meta name="fuzzy" content="maxDifference=0-255; totalPixels=208" />
+        <script>
+            if (window.internals) {
+                internals.settings.setEvaluationTimeZoomEnabled(false);
+            }
+        </script>
```
You don't need this, you can use `<!-- webkit-test-runner [ EvaluationTimeZoomEnabled=false ] -->` on top of your file instead.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/812859b3e0612332c8328ec3ea2f653d15865d5b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

weinig commented on Source/WebCore/style/values/backgrounds/StyleBorderImageWidth.h:
```diff
@@ -30,7 +30,7 @@
 namespace Style {
 
-struct BorderImageWidthValueLength : LengthWrapperBase<LengthPercentage<CSS::Nonnegative>> {
+struct BorderImageWidthValueLength : LengthWrapperBase<LengthPercentage<CSS::NonnegativeUnzoomed>> {
```
Be sure to mention this in the commit message, and add tests for it.

weinig commented on Source/WebCore/style/StyleResolver.cpp:
```diff
@@ -760,13 +760,15 @@ void Resolver::applyMatchedProperties(State& state, const MatchResult& matchResu
     if (builderState.style().usesViewportUnits())
         document().setHasStyleWithViewportUnits();
+
+    builderState.style().setUseSVGZoomRulesForLength(builderState.useSVGZoomRulesForLength());
```
Why do this after applying properties? Anything that might need the used value during style building would get things wrong.

nullhook replied:
this should be addressed here: https://github.com/WebKit/WebKit/pull/51643

weinig commented on Source/WebCore/rendering/style/StyleRareInheritedData.h:
```diff
@@ -230,7 +230,8 @@ class StyleRareInheritedData : public RefCounted<StyleRareInheritedData> {
     Style::BlockEllipsis blockEllipsis;
-
+    PREFERRED_TYPE(bool) unsigned enableEvaluationTimeZoom : 1;
+    PREFERRED_TYPE(bool) unsigned useSVGZoomRulesForLength : 1;
```
I can see why @anttik would suggest `enableEvaluationTimeZoom` be inherited (it is going to be the same for the every RenderStyle in a document) but I don't think you want `useSVGZoomRulesForLength` to be inherited, given that every RenderStyle needs this calculated independently based on the element.

weinig commented on Source/WebCore/rendering/style/CollapsedBorderValue.h:
```diff
@@ -39,7 +39,7 @@ class CollapsedBorderValue {
 
     CollapsedBorderValue(const BorderValue& border, const Color& color, BorderPrecedence precedence)
-        : m_width(border.nonZero() ? Style::evaluate<LayoutUnit>(border.width(), Style::ZoomNeeded { }) : 0_lu)
+        : m_width(border.nonZero() ? Style::evaluate<LayoutUnit>(border.width(), Style::ZoomFactor(1.0f)) : 0_lu)
```
Passing 1.0f is incorrect here.

weinig commented on Source/WebCore/rendering/RenderTableSection.cpp:
```diff
@@ -833,16 +848,22 @@ LayoutUnit RenderTableSection::calcInlineDirectionOuterBorder(InlineBorderSide s
-        if (cellBorder.style() > BorderStyle::Hidden && cellBorder.width() > borderWidth)
-            borderWidth = cellBorder.width();
+        if (cellBorder.style() > BorderStyle::Hidden) {
+            float cellBorderWidth = Style::evaluate<float>(cellBorder.width(), current.primaryCell()->style().usedZoomForLength());
```
I would pull `current.primaryCell()->style()` and `current.primaryCell()->parent()->style()` into variables up above where `cellBorder` and `rowBorder` are defined. Something like this:

```
        auto& cellBorderStyle = current.primaryCell()->style();
        const BorderValue& cellBorder = (side == InlineBorderSide::BorderStart) ? cellBorderStyle.borderStart(writingMode) : cellBorderStyle.borderEnd(writingMode);
        auto& rowBorderStyle = current.primaryCell()->parent()->style();
        const BorderValue& rowBorder = (side == InlineBorderSide::BorderStart) ? rowBorderStyle.borderStart(writingMode) : rowBorderStyle.borderEnd(writingMode);
```

That way it is super clear which style you are here (and below).

(I would apply this idea in the rest of the places in the file above as well).

weinig reviewed (changes requested):
Getting closer. 

I think the biggest thing would be getting more tests written for this. Especially the table code, where different zoom levels on different elements might render differently depending on zoom.

I think this would be more straightforward to review broken up a bit. I would suggest the following pieces.

1. Add `ZoomFactor` type and replace the use of explicit float zoom values. This would be a good change to introduce the new setting.
2. Convert just `-webkit-border-spacing`, adding new reference tests to ensure the changes to how zoom inheritance works, and showing that nothing else changes.
3. Convert `LineWidth` and `BorderImageWidth`, adding new reference tests to ensure the changes to how zoom inheritance works, and showing that nothing else changes.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d9ac36adb7757351b6df0d3b3799f3ea64101d21. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

nullhook commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -2691,6 +2691,20 @@ EnumeratingVisibleNetworkInterfacesEnabled:
+EvaluationTimeZoomEnabled:
+  category: css
+  type: bool
+  status: testable
```
it's not clear to me the reasoning behind marking this as "unstable" — how does one get visibility of all test failures when uploading a PR? this would require us to flip to testable, mark a per-test comment in header, and then flop it back to unstable.

nt1m replied:
Given we want to enable this relatively soon, I would prefer using `testable`.

nullhook replied:
it's unclear why this has to be marked as “unstable.” when uploading a PR, we need visibility into all test failures. however, that process can require flipping the status to “testable,” adding per-test comments in the header, and then reverting it back to “unstable.”

the issue becomes more problematic when other contributors join the effort of marking individual properties as “Unzoomed.” in this case, EWS will not raise any complaints, and PRs may be merged in an incomplete or inconsistent state.

brentfulgham replied:
My initial thought was to agree with @weinig  that we should mark this as `unstable` and opt individual tests in. However, after speaking with @nullhook in the hall, I better understand his position. As explained to me, there are two layers of opting in:
(a) A top layer (EvaluationTimeZoomEnabled), that controls the entire feature.
(b) Opting in the property to the new `Unzoomed` variant.

1. The flag being moved to `testable` does not activate new behavior until individual properties are moved to the new logic.
2. Presumably, a patch correcting a property (and switching e.g., `Nonnegative` to `NonnegativeUnzoomed`) would make sure tests pass under the new logic.

End users would not see the new behavior until we moved `EvaluationTimeZoomEnabled` from `testable` to `preview` (for STP) or `stable` (for all builds).

So in this case I think `testable` is the right state.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1efd6f771f185f491d57a3b4eb02fc164f3ffd53. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

nullhook commented:
PR for 1:

> Getting closer.
> 
> I think the biggest thing would be getting more tests written for this. Especially the table code, where different zoom levels on different elements might render differently depending on zoom.
> 
> I think this would be more straightforward to review broken up a bit. I would suggest the following pieces.
> 
> 1. Add `ZoomFactor` type and replace the use of explicit float zoom values. This would be a good change to introduce the new setting.
> 2. Convert just `-webkit-border-spacing`, adding new reference tests to ensure the changes to how zoom inheritance works, and showing that nothing else changes.
> 3. Convert `LineWidth` and `BorderImageWidth`, adding new reference tests to ensure the changes to how zoom inheritance works, and showing that nothing else changes.

1. https://github.com/WebKit/WebKit/pull/51643

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d799619d01f16f28485d6994056911fa97a8500a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fba34972c73a325ab3c4b92ebf18c6d779702c10. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

weinig commented:
My biggest concern here is all the test results that are changing. Please make sure to annotate the commit message with explanations as to why each new result is needed.

nullhook commented:
> My biggest concern here is all the test results that are changing.

This PR isn't ready yet. 

What is the "biggest concern" here? The test result changes that has been re-baselined are expected and necessary because this PR fundamentally changes how border-width behaves with zoom.

nullhook commented:
> The most basic form of zoom + border-width is not expected to change with this PR.

border-width on an element contributes to the visual space occupied by that element. if border-width is being transformed by zoom, then the visual space will also change accordingly, right?

nullhook commented:
> The most basic form of zoom + border-width is not expected to change with this PR.

border-width on an element contributes to the visual space occupied by that element. if border-width is being transformed by zoom, then the visual space will also change accordingly, right?

`<div style="border-left: 1px solid red; zoom: 2;">Hello</div>` 

the example does an identity transform (2*1) which effectively leaves the actual border value unchanged. 

however, if the following  `<div style="border-left: 2px solid red; zoom: 2;">Hello</div>` should effect the layout.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/90458978d1b6d1e15eadcace0b58e15e1716bd0c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f85483fee63053cc2167ce5b5a19eed70ad9502f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

nullhook commented on Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:
```diff
@@ -57,40 +57,39 @@ auto CSSValueConversion<LineWidth>::operator()(BuilderState& state, const CSSVal
 
-    // Any original result that was >= 1 should not be allowed to fall below 1. This keeps border lines from vanishing.
-
+    // For LineWidth, we need special logic to prevent lines from vanishing
```
this looks incomplete. you're missing the device scale factor logic. also, seems like this logic cannot be trivially moved over to evaluation-time as it requires access to doc's scale factor.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/17e2c21b8617a79b1ed12376bc8d850635d3866e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5da8a62a50f7981d5dd2de2e9cd9babc6d0b20d5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ac3dbef2831b918a3328fd33b9f6143b959406c1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9d5597af2911ac6c73d7b079cc0c8a12069c724c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

weinig commented:
> however, with the following  <div style="border-left: 2px solid red; zoom: 2;">Hello</div> should effect the layout.

Are you saying that the layout of this div will be different between before and after your change?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6587b6329dd6789ba9ae6d87fadb4fe325433786. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

nullhook commented:
> Are you saying that the layout of this div will be different between before and after your change?

no, i'm generalizing.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/83db05e1467ed38a5ddad9b1886c84a8fabdd1da. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51583

nullhook committed:
[CSS ZOOM] Apply zoom factor to border width
https://bugs.webkit.org/show_bug.cgi?id=299885
rdar://161669263

Reviewed by Antti Koivisto and Sam Weinig.

We apply the zoom factor at evaluation-time for LineWidth,
BorderImageWidth, and MaskBorderWidth.

Tests: imported/w3c/web-platform-tests/css/css-viewport/zoom/border-image-width-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/border-image-width.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/mask-border-width-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/mask-border-width.html

* LayoutTests/TestExpectations:

* LayoutTests/fast/sub-pixel/zoomed-em-border.html:
* LayoutTests/svg/zoom/page/zoom-replaced-intrinsic-ratio-001.htm:

We're disabling evaluation-time zoom flag on above tests because of double zoom.
The issue is that font-dependent units are being affected by the zoom
value during style-resolution, and we apply zoom again during evaluation.
This will be addressed separately as a follow-up:
https://bugs.webkit.org/show_bug.cgi?id=299816

* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-image-width-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-image-width-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-image-width.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-computed-zoom-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-computed-zoom.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-svg-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/mask-border-width-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/mask-border-width-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/mask-border-width.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/animations/border-width-interpolation-expected.txt:

A progression

* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
(WebCore::Layout::FormattingGeometry::computedBorder const):
* Source/WebCore/layout/formattingContexts/block/BlockFormattingGeometry.cpp:
(WebCore::Layout::BlockFormattingGeometry::intrinsicWidthConstraints const):
* Source/WebCore/layout/integration/LayoutIntegrationBoxGeometryUpdater.cpp:
(WebCore::LayoutIntegration::BoxGeometryUpdater::logicalBorder):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::fixedContainerEdges const):
* Source/WebCore/page/SpatialNavigation.cpp:
(WebCore::nodeRectInAbsoluteCoordinates):
* Source/WebCore/rendering/BorderEdge.cpp:
(WebCore::borderEdges):
(WebCore::borderEdgesForOutline):
* Source/WebCore/rendering/BorderPainter.cpp:
(WebCore::BorderPainter::paintOutline const):
* Source/WebCore/rendering/BorderShape.cpp:
(WebCore::BorderShape::shapeForBorderRect):
* Source/WebCore/rendering/NinePieceImagePainter.cpp:
(WebCore::computeSlice):
(WebCore::computeSlices):
(WebCore::paintNinePieceImage):
* Source/WebCore/rendering/RenderBoxModelObjectInlines.h:
(WebCore::RenderBoxModelObject::borderAfter const):
(WebCore::RenderBoxModelObject::borderBefore const):
(WebCore::RenderBoxModelObject::borderBottom const):
(WebCore::RenderBoxModelObject::borderEnd const):
(WebCore::RenderBoxModelObject::borderLeft const):
(WebCore::RenderBoxModelObject::borderRight const):
(WebCore::RenderBoxModelObject::borderStart const):
(WebCore::RenderBoxModelObject::borderTop const):
(WebCore::RenderBoxModelObject::borderWidths const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::drawFocusRing):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintAreaElementFocusRing):
* Source/WebCore/rendering/RenderMultiColumnSet.cpp:
(WebCore::RenderMultiColumnSet::paintColumnRules):
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::calcBorderStart const):
(WebCore::RenderTable::calcBorderEnd const):
(WebCore::RenderTable::outerBorderBefore const):
(WebCore::RenderTable::outerBorderAfter const):
(WebCore::RenderTable::outerBorderStart const):
(WebCore::RenderTable::outerBorderEnd const):
* Source/WebCore/rendering/RenderTableCell.cpp:
(WebCore::RenderTableCell::emptyBorder const):
(WebCore::RenderTableCell::computeCollapsedStartBorder const):
(WebCore::RenderTableCell::computeCollapsedEndBorder const):
(WebCore::RenderTableCell::computeCollapsedBeforeBorder const):
(WebCore::RenderTableCell::computeCollapsedAfterBorder const):
(WebCore::emptyBorder): Deleted.
* Source/WebCore/rendering/RenderTableCell.h:
* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::calcBlockDirectionOuterBorder const):
(WebCore::RenderTableSection::calcInlineDirectionOuterBorder const):
(WebCore::RenderTableSection::paintRowGroupBorderIfRequired):
(WebCore::RenderTableSection::cachedCollapsedBorder):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::extractControlStyleForRenderer const):
(WebCore::RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle const):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::paintMenuListButtonDecorationsForVectorBasedControls):
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::popupInternalPaddingBox const):
(WebCore::RenderThemeIOS::paintMenuListButtonDecorations):
* Source/WebCore/rendering/style/CollapsedBorderValue.h:
(WebCore::CollapsedBorderValue::CollapsedBorderValue):
(WebCore::CollapsedBorderValue::m_transparent): Deleted.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setDeviceScaleFactor):
(WebCore::RenderStyle::imageOutsets const):
(WebCore::RenderStyle::outlineWidth const):
(WebCore::RenderStyle::outlineOffset const):
(WebCore::RenderStyle::outlineSize const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::deviceScaleFactor const):
(WebCore::RenderStyle::usedZoomForLength const):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::computeFloatVisibleRectInContainer):
* Source/WebCore/style/StyleResolveForDocument.cpp:
(WebCore::Style::resolveForDocument):
* Source/WebCore/style/values/backgrounds/StyleBorderImageWidth.h:
* Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:
(WebCore::Style::CSSValueConversion<LineWidth>::operator):
(WebCore::Style::FloatBoxExtent>::operator):
(WebCore::Style::LayoutBoxExtent>::operator):
* Source/WebCore/style/values/backgrounds/StyleLineWidth.h:
(WebCore::Style::LineWidth::LineWidth):
* Source/WebCore/style/values/masking/StyleMaskBorderWidth.h:
* Source/WebCore/style/values/primitives/StyleZoomPrimitives.h:
(WebCore::Style::ZoomFactor::ZoomFactor):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::absoluteInteractionBounds):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMNode innerFrameQuad]):

Canonical link: https://commits.webkit.org/301536@main
